### PR TITLE
Feature/ja/learning fetch update/nm

### LINKED ISSUE #732: API: Learning update to support calculations
Ticket: https://trello.com/c/nmWCrCAP/251-learning-distribution-to-be-based-on-time-per-type-not-the-count-per-type

Update to `/me/learnings/` to support calculations for '**Learning Distribution**' and '**Learning goal**'

# Response 

Current response 
```lang-js
[
  {
    "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
    "name": "string",
    "duration_minutes": 32767,
    "date_completed": "2022-11-09",
    "cost_pounds": 32767,
    "cost_unknown": true,
    "learning_type": "On the job"
  }
]
```

New response
Similar to the 'reports' work we did. Can you replace the original with this
```lang-js
{
  "distribution": [
    { "name": "Formal", "value_percentage": 40 },
    { "name": "Social", "value_percentage": 40 }
    { "name": "On the job", "value_percentage": 20 }
  ],
  "goal_value_days": 1,
  "goal_value_percentage": 7,
  
  "data": [ //The learning records from the original response is moved here
    {
      "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
      "name": "string",
      "duration_minutes": 32767,
      "date_completed": "2022-11-09",
      "cost_pounds": 32767,
      "cost_unknown": true,
      "learning_type": "On the job"
    }
  ]
}
```

This will help me to produce the following on UI 

Screenshot 2022-11-09 at 17 27 49

# Parameters

Keep all the existing params like 'sorting' the same

# Branch

This will be a breaking change on Ellandi. Therefore can we work on the same branch for this so this code can be released the same time 

Branch 
```
feature/ja/learning-fetchupdate

### PR DESCRIPTION
**Pull request by nmenezes0**
Fixes #732. Don't merge into develop til @rottitime's frontend code is ready (otherwise it will break) - this is why PR marked as draft, even though it can be reviewed. @rottitime - I took a branch off your branch `feature/ja/learning_fetch_update/`.

This makes changes to the `/me/learning/` endpoints to add aggregate data for reporting.

We were unsure of what behaviour should be when we pass in a `learning_type` parameter to the `/me/learning/` endpoint - and decided it should return summary stats for all learning (as otherwise, they wouldn't really make sense), and then data for learning of that type (as before).

### CODE REVIEW

File: api/ellandi/learning.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+from django.db.models import Avg, Sum
+
+from ellandi.registration.models import Learning
+
+HOURS_IN_WORK_DAY = 7.4
+LEARNING_TARGET_IN_DAYS = 10
+MINUTES_IN_LEARNING_TARGET = LEARNING_TARGET_IN_DAYS * HOURS_IN_WORK_DAY * 60
+
+
+def get_start_financial_year():
```
Comment (Singletoned) [inline api/ellandi/learning.py:12]: It's almost always best not to try to work these things out yourself; there's lots of weird edge cases.  Use a relativedelta from https://dateutil.readthedocs.io/en/stable/relativedelta.html

File: api/ellandi/learning.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    learning_qs = Learning.objects.filter(user__in=users_qs).filter(date_completed__gte=start_financial_year)
+    return learning_qs
+
+
+def get_summary_course_costs(learning_qs):
+    formal_learning_with_costs = learning_qs.filter(learning_type="Formal").exclude(cost_unknown=True)
+    cost_aggregates = formal_learning_with_costs.aggregate(Sum("cost_pounds"), Avg("cost_pounds"))
+    avg_cost = cost_aggregates["cost_pounds__avg"] or 0
+    sum_cost = cost_aggregates["cost_pounds__sum"] or 0
+    return round(avg_cost), round(sum_cost)
```
Comment (Singletoned) [inline api/ellandi/learning.py:32]: Always leave `return` statements bare unless it's a one liner.  It's really easy for people to miss that something is happening there

File: api/ellandi/learning.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def get_total_avg_learning_financial_year(learning_qs, total_users):
+    if not total_users:
+        return 0, 0
+    total_learning = learning_qs.aggregate(Sum("duration_minutes"))["duration_minutes__sum"]
+    total_learning = total_learning or 0
+    avg_learning_mins = total_learning / total_users
+    proportion_learning_per_user = avg_learning_mins * 100 / MINUTES_IN_LEARNING_TARGET
+    days_per_user = round(avg_learning_mins / (60 * HOURS_IN_WORK_DAY))
```
Comment (Singletoned) [inline api/ellandi/learning.py:71]: Tiny thing, but `HOURS * 60` makes more sense than `60 * HOURS`

**Review by Singletoned — CHANGES_REQUESTED** [review]

File: api/ellandi/learning.py (hunk 2 above)
Comment (nmenezes0) [reply api/ellandi/learning.py:32]: By bare do you mean do the calcs like `round` earlier - and not in the return statement?

File: api/ellandi/learning.py (hunk 1 above)
Comment (nmenezes0) [reply api/ellandi/learning.py:12]: I'm a bit confused as to what bit you expect me to use the relativedelta for?

File: api/ellandi/learning.py (hunk 2 above)
Comment (nmenezes0) [reply api/ellandi/learning.py]: I assumed that's what you meant and made that change.

File: api/ellandi/learning.py (hunk 3 above)
Comment (nmenezes0) [reply api/ellandi/learning.py]: have changed

File: api/ellandi/learning.py (hunk 1 above)
Comment (nmenezes0) [reply api/ellandi/learning.py:13]: Have changed - have a look if you can see you're happy with my changes.

File: api/ellandi/learning.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+from ellandi.registration.models import Learning
+
+HOURS_IN_WORK_DAY = 7.4
+LEARNING_TARGET_IN_DAYS = 10
+MINUTES_IN_LEARNING_TARGET = LEARNING_TARGET_IN_DAYS * HOURS_IN_WORK_DAY * 60
+
+
+def get_start_financial_year():
+    today = datetime.date.today()
+    if today.month < 4:
```
Comment (Singletoned) [inline api/ellandi/learning.py:15]: You don't need to do this anymore. `start_financial_year = today - relativedelta.relativedelta(month=4, day=1)` should always give the right answer

**Review by Singletoned — CHANGES_REQUESTED** [review]

File: api/ellandi/learning.py (hunk 4 above)
Comment (nmenezes0) [reply api/ellandi/learning.py:15]: It did not give the right answer:

```
In [7]: date(2022, 1, 1) - relativedelta.relativedelta(month=4, day=1)
Out[7]: datetime.date(2022, 4, 1)
```

Comment (Singletoned) [reply api/ellandi/learning.py:15]: Use https://pypi.org/project/fiscalyear/ instead then

**Pull request merged**